File: python/zebra-puzzle/zebra_puzzle.py
# coding=utf-8
from __future__ import unicode_literals
import itertools

__author__ = 'Cedric Zhuang'

PEOPLE = ['Norwegian', 'Englishman', 'Spaniard', 'Ukrainian', 'Japanese']


def next_to(x, y):
    return abs(x - y) == 1


def right_to(x, y):
    return x - 1 == y


def solution():
    houses = first, _, middle, _, _ = range(5)
    p = list(itertools.permutations(houses))
    ans = next(
        (PEOPLE[water], PEOPLE[zebra])
        for (Norwegian, Englishman, Spaniard, Ukrainian, Japanese) in p
        if Norwegian == first
        for (green, red, ivory, yellow, blue) in p
        if Englishman == red and
        right_to(green, ivory) and
        next_to(Norwegian, blue)
        for (dog, snails, fox, horse, zebra) in p
        if Spaniard == dog
        for (coffee, tea, milk, orange_juice, water) in p
        if coffee == green and
        Ukrainian == tea and
        milk == middle
        for (old_gold, kools, chesterfields, lucky_strike, parliaments) in p
        if old_gold == snails and
        kools == yellow and
        next_to(chesterfields, fox) and
        next_to(kools, horse) and
        lucky_strike == orange_juice and
        Japanese == parliaments)

    return ("It is the {} who drinks the water.\n"
            "The {} keeps the zebra.").format(ans[0], ans[1])

File: scala/zebra-puzzle/src/main/scala/ZebraPuzzle.scala
object ZebraPuzzle extends Enumeration {
  type Resident = Value
  val Englishman, Spaniard, Ukrainian, Norwegian, Japanese, Unknown = Value

  case class Solution(waterDrinker: Resident, zebraOwner: Resident)

  private def p: Iterator[Seq[Int]] = (0 to 4).permutations

  private def isRightTo(x: Int, y: Int): Boolean = x - y == 1

  private def nextTo(x: Int, y: Int): Boolean = math.abs(x - y) == 1

  private val Seq(first: Int, _, middle: Int, _, _) = 0 to 4

  lazy val solve: Solution = solutions.next()

  private def solutions: Iterator[Solution] =
    for (residents <- p;

         Seq(englishman: Int, spaniard: Int, ukrainian: Int,
         norwegian: Int, japanese: Int) <- Seq(residents)
         if norwegian == first;

         Seq(red: Int, blue: Int, green: Int, ivory: Int, yellow) <- p
         if englishman == red
           && isRightTo(green, ivory)
           && nextTo(norwegian, blue);

         Seq(dog: Int, snail: Int, zebra: Int, fox: Int, horse: Int) <- p
         if spaniard == dog;

         Seq(coffee, tea: Int, milk: Int, orangeJuice: Int, water: Int) <- p
         if green == coffee
           && tea == ukrainian
           && milk == middle;

         Seq(oldGold: Int, kools: Int, chesterfields: Int,
         luckyStrike: Int, parliaments: Int) <- p
         if oldGold == snail
           && kools == yellow
           && nextTo(chesterfields, fox)
           && nextTo(kools, horse)
           && luckyStrike == orangeJuice
           && japanese == parliaments)

      yield Solution(
        getResident(water, residents),
        getResident(zebra, residents))

  private def getResident(house: Int, residents: Seq[Int]) =
    ZebraPuzzle(residents.zipWithIndex.find { case (v, _) => v == house }.get._2)
}

